pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prober
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/probe"
+	httpprobe "k8s.io/kubernetes/pkg/probe/http"
+	"net/http"
+	"time"
+)
+
+// probeRunner is an interface for running probes, objects implementing
+// probeRunner encapsulates the logic of running probes.
+type probeRunner interface {
+
+	// init initializes objects, which are later reused by run on every probe.
+	init(probeSpec v1.Probe, container v1.Container, status v1.PodStatus) error
+
+	// run holds the logic of running probe, run takes prober as a dependency
+	// to run the probe on either http.Prober or tcp.Prober or grpc.Prober or exec.Prober
+	run(ctx context.Context, prober prober) (probe.Result, string, error)
+}
+
+// tcpProbeRunner holds state and logic for running TCP probes.
+type tcpProbeRunner struct {
+	host    string
+	port    int
+	timeout time.Duration
+}
+
+// newTCPProbeRunner returns tcpProbeRunner which implements probeRunner
+func newTCPProbeRunner() *tcpProbeRunner {
+	return &tcpProbeRunner{}
+}
+
+func (tp *tcpProbeRunner) init(probeSpec v1.Probe, container v1.Container, status v1.PodStatus) error {
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+	port, err := probe.ResolveContainerPort(probeSpec.TCPSocket.Port, &container)
+	if err != nil {
+		return err
+	}
+
+	host := probeSpec.TCPSocket.Host
+	if host == "" {
+		host = status.PodIP
```

Comment:
@aojea do you think PodIP can change?

@daman1807 I think overriding the PodIP will need to be implemented in `run` method. IP can change

Reply:
Same for both - http and grpc

Reply:
@SergeyKanzhelev @aojea If PodIP changes only when the pod is restarted or recreated on some other node, then the worker already handles this case. 

https://github.com/kubernetes/kubernetes/blob/c9bbb6553d11350f37192c16f28825854b020630/pkg/kubelet/prober/worker.go#L215-L220

Reply:
yes, I understand this part. I didn't check if there is any codepath that will update the PodIP on the running Pod. I don't think probes may start before IP was assigned. But maybe containerd got restarted or something like this. I heavily doubt it, but @aojea mentioned this so I wanted to be on the safe side.

If you can do this analysis, please add it to PR so we can be sure it can be cached.

Reply:
one thing is the current behavior of Kubelet, and other is the pod.Status.PodAPI API semantics, since pod.Status.PodIP is mutable I feel risky to do this change based on kubelet existing behavior-

Per example, there are conversations in MultiNetwork where a Pod may not use CNI to get the IP addresses, I don't know how likely is this to happen and when it can happen, but it looks like a possibility to me and this will break or it will cause these new developments to adapt to an implementation detail like this.

Let me tag @thockin @liggitt and @smarterclayton , we touched a bit on this here https://github.com/kubernetes/kubernetes/issues/109414#issuecomment-1125160058 and I'd like to have a wider consensus before we commit to this, probes are a critical part of the system.

Reply:
@SergeyKanzhelev @aojea 
I'm now tracking podIP at probe worker and setting up probeRunner again if the IP changes.